### ISSUE TITLE
Variant on components

### ISSUE THREAD

**Issue by bulby97** [chakra-ui/chakra-ui#435]
At the moment, this is the best way I found to render components based on state: 

```
const useStyles = ({ variant }) => ({
  fontWeight: 'bold',
  py: 2,
  px: 4,
  rounded: 'full',
  fontSize: 'xs',
  d: 'inline-block',

  ...variant === 'full' && {
    bg: 'tomato',
    color: 'white',
  },

 ...variant === 'outlined' && {
    bg: 'tomato',
    color: 'white',
  },

});
```

```
  const stylesProps = useStyles(props)
  return <Box {...stylesProps} />
```

Is this a good pattern  ? Is there any better solution ?

**Comment by bulby97** [comment]
Thinking about it, maybe we're doing it backwards ?
https://theme-ui.com/sx-prop has a clever way of doing it : stying inline, using values from the theme.

Truth is: 70% of components we're creating need to be responsive. And some will not have just two variants, but sometimes much more.

**Comment by segunadebayo** [comment]
Hi @bulby97,

Your solution works for now. I understand this is a bit painful to handle variants that way.

We're solved this permanently in the next release. Let me give you a glimpse:
1. You will be able to use the `sx`prop in any component
2. You can define a component schema to support different variants, sizes and color schemes. 

```tsx
// step 1: define component schema in theme.components
const theme = {
  components: {
    MyComponent: {
      // you can define variants
      variants: {
        outlined: {}, // your style props
        full: {}, // your style props
      },
      // you can also define sizes
      sizes: {
        small: {}, // your style props
        big: {}, // your style props
      },
    },
  },
}

// step 2: create a chakra component
const MyComponent = createChakra("button", { themeKey: "MyComponent" })

// step 3: Just use the component :)
<MyComponent variant="outlined" size="small"/>
```

**Comment by christofferberg** [comment]
@segunadebayo 

Is it possible to tell when we can expect the next release? 🆒

**Comment by segunadebayo** [comment]
@christofferberg, I'd say by the end of March.

**Comment by bulby97** [comment]
This is amazing! Thank you for your hard work.

One question though : what is the purpose of declaring variants inside theme? I know it's used in theme-ui, but will it not create at some point different sources of truth?

**Comment by segunadebayo** [comment]
@bulby97, it's mostly to have one source of truth for component styles and variants.

This means you can simply import any Chakra components, and override their "global" style from the theme or even make your own styles.

It improves the themeability experience, and ensure all apps built with Chakra don't look the same 😀

**Comment by glenmcrae** [comment]
Great work @segunadebayo 👍

**Comment by lpysj** [comment]
Is this available in the current release? (Either the sx prop or the component key)

**Comment by with-heart** [comment]
@lpysj They are not, but they're coming quite soon. #696

**Comment by lpysj** [comment]
@with-heart Thank you :)

**Comment by with-heart** [comment]
This is now part of the component theming api available in the 1.0 alpha! You can explore the docs for this feature at [next.chakra-ui.com](https://next.chakra-ui.com/docs/theming/overview) under the "Theming" category.

**Comment by ankursehdev** [comment]
Is there documentation of this? I am trying to do this but variants are not working

```
const InformationTextContainer: ChakraComponent<'div'> = chakra(Box, {
  baseStyle: () => ({
    display: 'flex',
    border: '3px solid blue',
  }),
  variants: {
    primary: ({ theme }) => ({
      ...theme.menuDefaultButton,
    }),
    secondary: ({ theme }) => ({
      border: '1px solid black',
      ...theme.informationtextDefaultText,
    }),
  },
  defaultProps: {
    variant: 'secondary',
  },
});

--
<InformationTextContainer>Text here</InformationTextContainer>
```

This doesn't work! Any pointers what I am doing wrong?
